### PR TITLE
Implemented Issue from Issue #9

### LINKED ISSUE #9: [Backend] Issues
As a user, I should be able to submit a new issue into Ticket Penguin, view existing issues, update an issue, or delete an issue. 

- New issues should be submitted to the `/organizations/{id}/issues` endpoint via HTTP POST request and persisted to the database.
- Existing issues should be returned when an HTTP GET request is sent to the `/organizations/{id}/issues` endpoint.
- An issue may be updated via HTTP PATCH request at the `/organizations/{org_id}/issues/{issue_id}` endpoint. Only fields included in the PATCH request body should be updated.
- An issue may be deleted via HTTP DELETE request at the `/organizations/{org_id}/issues/{issue_id}` endpoint.

NOTE: Special care should be taken to confirm that the requesting user has access to the organization specified.

### PR DESCRIPTION
**Pull request by Howllo**
This implements the ability to create, update, get, and delete issues. User wills be able to create new issue only if there is a column that allows the issues to be assigned to it.

### CODE REVIEW

**Review by nathanaelg16 — CHANGES_REQUESTED** [review]
The issues should be assigned to a particular organization, and checks should be performed to make sure the user belongs to the organization where the issue is being submitted. I implemented organizations in my [PR](https://github.com/Devature/penguin-api/pull/29).

File: src/main/java/dev/devature/penguin_api/controller/IssueController.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/controller/IssueController.java]: We can do a PostMapping to the base issue URL here

```suggestion
    @PostMapping
```

File: src/main/java/dev/devature/penguin_api/controller/IssueController.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/controller/IssueController.java:65]: We should also do a `@GetMapping` on the base issue URL that returns all active issues for the User's organization

File: src/main/java/dev/devature/penguin_api/service/IssueService.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/service/IssueService.java]: [HTTP PATCH](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH) requests can do partial modifications of the resource, whereas HTTP PUT does a complete replacement of the resource. It seems like you are intending this to be for a PATCH request, so this method should take special care in determining which fields are being updated. More specifically, the incoming JSON request might only contain a subset of the fields that make up an Issue, so the Issue object will have the missing fields as null. In this case, we only want to update the non-null fields and keep the rest intact.

File: src/main/java/dev/devature/penguin_api/entity/Member.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/entity/Member.java:1]: We actually don't need this entity since it's created implicitly by the ManyToMany relationship users have with organizations. See [here](https://github.com/Devature/penguin-api/blob/74e021165e31495cdd5bc08497607021ecf1ed80/src/main/java/dev/devature/penguin_api/entity/User.java#L55) and [here](https://github.com/Devature/penguin-api/blob/74e021165e31495cdd5bc08497607021ecf1ed80/src/main/java/dev/devature/penguin_api/entity/Organization.java#L25).

File: src/main/java/dev/devature/penguin_api/service/IssueService.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/service/IssueService.java]: A user is authorized to create an issue if they are a member of the organization the issue is being created in. So we should check if a user is a member similar to what's done [here](https://github.com/Devature/penguin-api/blob/74e021165e31495cdd5bc08497607021ecf1ed80/src/main/java/dev/devature/penguin_api/service/OrganizationService.java#L48).

File: src/main/java/dev/devature/penguin_api/repository/OrganizationRepository.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/repository/OrganizationRepository.java]: ~~Since the permissions check involves checking whether a user is a member of the organization, this is no longer needed.~~ Could you explain why we're querying the Organization table to find if the user is a member? I figured this info would be in the `Members` table...

File: src/main/java/dev/devature/penguin_api/service/IssueService.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/service/IssueService.java:32]: One major flaw with this flow is that it assumes the issue's `createdByID` is supplied by the API caller, and that the API caller isn't lying to us about who they are. As it stands, this flow is susceptible to a bug where user Alice calls the API to create a new issue, but supplies user Bob's user ID instead. So the Issue is created by Alice, but Bob is listed as the creator. Instead, we should have the API automatically set the calling user as the creator. We can get this information from the auth token, since this will uniquely identify the user who is calling our API. Check the flow I used for the Organizations endpoint. [This](https://github.com/Devature/penguin-api/blob/74e021165e31495cdd5bc08497607021ecf1ed80/src/main/java/dev/devature/penguin_api/service/OrganizationService.java#L52) is how I got the calling user's email address, and [this](https://github.com/Devature/penguin-api/blob/74e021165e31495cdd5bc08497607021ecf1ed80/src/main/java/dev/devature/penguin_api/controller/OrganizationController.java#L39) is how to get the claims from the auth token.

File: src/main/java/dev/devature/penguin_api/service/IssueService.java
Comment (nathanaelg16) [inline src/main/java/dev/devature/penguin_api/service/IssueService.java:1]: The authorization check should be done for each of the CRUD methods here. Only users who are members of the issue's organization should be allowed to create, get, update, or delete an issue. (This is only a problem since we're having a shared schema with other orgs. We want to try to prevent a rogue user from deleting or changing issues for an organization they are not a part of.)

**Review by nathanaelg16 — CHANGES_REQUESTED** [review]
Thanks for your hard work on this! I left a few comments on things that need to be addressed before we can merge this.

File: src/main/java/dev/devature/penguin_api/entity/Member.java
Comment (Howllo) [reply src/main/java/dev/devature/penguin_api/entity/Member.java:1]: When designing this system, I didn't intend for the User entity to handle organization-specific data. Doing so would introduce unnecessary bloat to the User table without adding any value and would break normalization principles. The purpose of the User entity is solely for authentication, while the Member entity is responsible for authorization within an organization.

- Normalization: The User table should remain focused on globally unique identity data, such as login credentials, whereas the Member entity handles organization-specific details (e.g., roles, ranks, and status).
- Flexibility: Since a User can belong to multiple organizations, storing organization-specific data in the User table would lead to scaling issues and data redundancy. By placing this information in the Member table, we keep things more efficient and scalable.

A simple analogy would be that the User entity functions like a driver's license—it authenticates the person across the system, while the Member entity is like a company badge, authorizing their actions within an organization. This separation ensures a clean, maintainable design.

File: src/main/java/dev/devature/penguin_api/controller/IssueController.java
Comment (Howllo) [reply src/main/java/dev/devature/penguin_api/controller/IssueController.java:65]: The suggestion to return all active issues for the User's organization doesn't align with the current design. In this system, issue management is best handled by the columns, which represent different stages or statuses within a project. Since an organization can have multiple projects, and each project may have its own set of issues, retrieving all issues for the organization would be inefficient and unnecessary.

The IssueController should be responsible solely for CUD (Create, Update, and Delete) operations, while the columns handle the propagation and retrieval of issues based on their status. As such, I will remove the getIssue method, as it doesn't serve a purpose in this design, and delegate issue retrieval to the service layer instead.

File: src/main/java/dev/devature/penguin_api/entity/Member.java
Comment (nathanaelg16) [reply src/main/java/dev/devature/penguin_api/entity/Member.java:1]: So this doesn't actually add any organization data to the User table. Since we are using an ORM, by specifying a many to many relationship with the Organizations table, the ORM will automatically create a new table to handle this relationship, in this case the table is called `organization_members`. The User table remains intact.

However, it seems like you intend to include extra fields in this member table (like roles, ranks, etc.), and I don't think the ORM is equipped to handle that this way, so I guess this is fine. I will fix and update my code to reflect this new Members entity after we merge this PR.

File: src/main/java/dev/devature/penguin_api/service/IssueService.java
Comment (nathanaelg16) [reply src/main/java/dev/devature/penguin_api/service/IssueService.java]: Disregard that link, that code will need to be changed. However we still need to check if the member belongs to the organization, which I don't think your call to the `OrganizationRepository` does well. Is there something I'm missing?

File: src/main/java/dev/devature/penguin_api/controller/IssueController.java
Comment (nathanaelg16) [reply src/main/java/dev/devature/penguin_api/controller/IssueController.java:65]: Fair point. I think you should still leave in the `getIssue` method you've written though, as it would be useful to link to and retrieve a singular issue. But if you feel strongly towards removing it, I'm fine with that, too.

**Review by nathanaelg16 — APPROVED** [review]

**Pull request merged**
